text/0013-prepare-blockbuilder-and-core-runtime-apis-for-mbms.md

```diff
@@ -0,0 +1,126 @@
+# RFC-0013: Prepare `BlockBuilder` and `Core` runtime APIs for MBMs
+
+|                 |                                                                             |
+| --------------- | --------------------------------------------------------------------------- |
+| **Start Date**  | July 24, 2023 |
+| **Description** | Prepare the `BlockBuilder` and `Core` Runtime APIs for Multi-Block-Migrations. |
+| **Authors**     | Oliver Tale-Yazdi |
+
+## Summary
+
+Introduces breaking changes to the `BlockBuilder` and `Core` runtime APIs.  
+A new function `BlockBuilder::last_inherent` is introduced and the return value of `Core::initialize_block` is changed to an enum.  
+The versions of both APIs are bumped; `BlockBuilder` to 7 and `Core` to 5.
+
+## Motivation
+
+There are three main features that motivate for this RFC:
+1. Multi-Block-Migrations: These make it possible to split a migration over multiple blocks.
+2. Pallet `poll` hook: Can be used to gradually replace `on_initialize`/`on_finalize` in places where the code does not need to run by a hard deadline, since it is not guaranteed to execute each block.
+3. New callback `System::PostInherents`: Can replace `on_initialize`/`on_finalize` where a hard deadline is required (complements `poll`). It is guaranteed to execute each block.
+
+These three features can be implemented when fulfilling these two requirements:
+1. The runtime can tell the block author to not include any transactions in the block.
+2. The runtime can execute logic right after all pallet-provided inherents have been applied.
+
+## Stakeholders
+
+- Substrate Maintainers: They have to implement this, including tests, audit and
+  maintenance burden.
+- Polkadot Runtime developers: They will have to adapt the runtime files to this breaking change.
+- Polkadot Parachain Teams: They also have to adapt to the breaking changes but then eventually have
+  multi-block migrations available.
+
+## Explanation
+
+
+### `Core::initialize_block`
+
+This runtime API function is changed from returning `()` to `ExtrinsicInclusionMode`:
+```rust
+enum ExtrinsicInclusionMode {
+  /// All extrinsics are allowed in this block.
+  AllExtrinsics,
+  /// Only inherents are allowed in this block.
+  OnlyInherents,
+}
+```
+
+A block author MUST respect the `ExtrinsicInclusionMode` that is returned by `initialize_block`. The runtime MAY reject blocks that violate this requirement. 
```

Comment:
```suggestion
A block author SHOULD respect the `ExtrinsicInclusionMode` that is returned by `initialize_block`. The runtime will reject any non-inherent transaction, meaning the block author would only waste computation time. 
```

Reply:
IMO "MUST" is better.
The runtime isn't meant to try to detect bugs in the host, because it can't anyway. It can only assume that the host behaves as expected. In that specific situation, the runtime is able to detect bugs in the host, but I don't see why we should create an exception here just because it's possible.
I'd even suggest that the runtime panics if the host ignores the `ExtrinsicInclusionMode`.

Reply:
> I'd even suggest that the runtime panics if the host ignores the `ExtrinsicInclusionMode`.

Yeah 100%. This is what "will reject" means.

Reply:
I was using `MAY` since the inherent Extrinsic could have been a storage No-OP, in which case the runtime would have no way of detecting it.  
But it is probably easier to use `MUST` here.